### DISCUSSION TITLE
Inconsistent results when copying pytorch model weights to flax model

### DISCUSSION THREAD

**Discussion by ruili3** [google/flax#2997] [Q&A]
### Problem statement
I tried to copy Pytorch ResNet18 weights to Flax ResNet18 following the [guidance](https://flax.readthedocs.io/en/latest/advanced_topics/convert_pytorch_to_flax.html#:~:text=%2C%20t_out).  However, I encountered inconsistency results even from the first conv layer. The problem seems to occur in the default padding setting in Flax. I provide minimum code to reproduce it. 
```
import jax
import jax.numpy as jnp
import numpy as np
from functools import partial
import os
import flax
import torch
from flax import linen as nn
from jax.tree_util import tree_flatten
from clu import parameter_overview

def to_numpy(x):
  return x.detach().cpu().numpy()


from flax.linen.initializers import variance_scaling

input_color = np.load("/path/to/input/tensor/pseudo_input.npy")
print(input_color.mean())
print(input_color.shape)
pyt_enc_dict = torch.load("/path/to/pytorch/model/encoder.pth")

wt = to_numpy(pyt_enc_dict["encoder.conv1.weight"])

pyt_conv = torch.nn.Conv2d(3, 64, kernel_size=7, stride=2, padding=3, bias=False)
jax_conv = nn.Conv(
    features=64,
    kernel_size=(7, 7),
    strides=(2, 2),
    padding= 'SAME',
    use_bias=False,
    name="jax_conv")

variables = {'params': {'kernel': jnp.transpose(wt, (2, 3, 1, 0))}}
pyt_conv.weight = torch.nn.Parameter(torch.from_numpy(wt))

py_out = pyt_conv(torch.from_numpy(input_color).permute(0,3,1,2))
py_out = to_numpy(py_out.permute(0,2,3,1))

jax_out = jax_conv.apply(variables, input_color)

np.testing.assert_almost_equal(py_out, jax_out, decimal=6)
```
where the "/path/to/input/tensor/pseudo_input.npy" is generated using 
`key = jax.random.PRNGKey(0), x = jax.random.normal(key, (1, 192, 640, 3))` and "/path/to/pytorch/model/encoder.pth" points to a trained PyTorch ResNet18 model.
After running the code, the resulting tensors are inconsistent in over 90% of elements. 

> Arrays are not almost equal to 6 decimals
> 
> Mismatched elements: 1777998 / 1966080 (90.4%)
> Max absolute difference: 21.563568
> Max relative difference: 2943599.
>  x: array([[[[ 1.013199e+00,  9.168627e-01, -1.924659e-06, ...,
>           -1.139887e+00, -3.485720e-01, -1.048833e+00],
>          [ 4.321716e+00,  1.305295e+00, -1.952268e-06, ...,...
>  y: array([[[[-4.782545e+00,  1.486091e+00, -1.172088e-06, ...,
>           -1.071571e+00, -1.331989e-01, -4.518164e-01],
>          [-1.766392e+00,  1.878799e+00, -1.992945e-06, ...,...

But when I change part of the code as follows, the inconsistency cases are largely reduced.
```
padding_size = (7 - 1) // 2  # 3

pyt_conv = torch.nn.Conv2d(3, 64, kernel_size=7, stride=2, padding=3, bias=False)
jax_conv = nn.Conv(
    features=64,
    kernel_size=(7, 7),
    strides=(2, 2),
    padding=((padding_size, padding_size), (padding_size, padding_size)),
    use_bias=False,
    name="jax_conv")
```

> Arrays are not almost equal to 6 decimals
> 
> Mismatched elements: 14614 / 1966080 (0.743%)
> Max absolute difference: 6.67572e-06
> Max relative difference: 0.06328507
>  x: array([[[[ 1.013199e+00,  9.168627e-01, -1.924659e-06, ...,
>           -1.139887e+00, -3.485720e-01, -1.048833e+00],
>          [ 4.321716e+00,  1.305295e+00, -1.952268e-06, ...,...
>  y: array([[[[ 1.013199e+00,  9.168626e-01, -1.924659e-06, ...,
>           -1.139887e+00, -3.485718e-01, -1.048833e+00],
>          [ 4.321715e+00,  1.305295e+00, -1.952268e-06, ...,...

### My question

- I'm curious about why it can happen. Because setting `padding='SAME'` should act the same as `padding=((padding_size, padding_size), (padding_size, padding_size))`.
- Even after changing the code to `padding=((padding_size, padding_size), (padding_size, padding_size))`, there are still tensor elements that are not consistent up to 6 decimals. Are there still any issues to address in this code?
Thank you a lot for the answer.

### System information
- OS Platform and Distribution (Linux Ubuntu):
- Flax, jax, jaxlib versions:
> Name: flax
> Version: 0.4.2
> Summary: Flax: A neural network library for JAX designed for flexibility
> Home-page: https://github.com/google/flax
> Author: Flax team
> Author-email: flax-dev@google.com
> License: UNKNOWN
> Requires: jax, matplotlib, msgpack, numpy, optax, typing-extensions
> Required-by: clu, orbax
> 
> Name: jax
> Version: 0.3.24
> Summary: Differentiate, compile, and transform Numpy code.
> Home-page: https://github.com/google/jax
> Author: JAX team
> Author-email: jax-dev@google.com
> License: Apache-2.0
> Requires: numpy, opt-einsum, scipy, typing-extensions
> Required-by: chex, clu, flax, optax, orbax
> 
> Name: jaxlib
> Version: 0.3.24+cuda11.cudnn82
> Summary: XLA library for JAX
> Home-page: https://github.com/google/jax
> Author: JAX team
> Author-email: jax-dev@google.com
> License: Apache-2.0
> Requires: numpy, scipy
> Required-by: chex, clu, optax, orbax

- Python version: 3.8.5

**Comment by cgarciae** [comment]
Hey @ruili3, converted this to a discussion because strict pytorch compatibility is not a goal for Flax, the guide is there for convenience. That said, we can try figure out the differences here :)

**Comment by cgarciae** [comment]
Looking at the [jax-resnet](https://github.com/n2cholas/jax-resnet/blob/main/jax_resnet/resnet.py#L25-L29) repo, it seems that you have to explicitly set the padding to `[(3, 3), (3, 3)]` to match pytorch implementation:

```python

import jax
import jax.numpy as jnp
import numpy as np
import torch

from flax import linen as nn


def to_numpy(x):
  return x.detach().cpu().numpy()


pyt_conv = torch.nn.Conv2d(3, 64, kernel_size=7, stride=2, padding=3, bias=False)
jax_conv = nn.Conv(
    features=64,
    kernel_size=(7, 7),
    strides=(2, 2),
    # padding= 'SAME',
    padding=[(3, 3), (3, 3)],
    use_bias=False,
    name="jax_conv")

key = jax.random.PRNGKey(0)
x = np.asarray(jax.random.normal(key, (1, 192, 640, 3)))
wt = to_numpy(pyt_conv.weight)
variables = {'params': {'kernel': jnp.transpose(wt, (2, 3, 1, 0))}}
pyt_conv.weight = torch.nn.Parameter(torch.from_numpy(wt))

py_out = pyt_conv(torch.from_numpy(x).permute(0,3,1,2))
py_out = to_numpy(py_out.permute(0,2,3,1))

jax_out = jax_conv.apply(variables, x)

np.testing.assert_almost_equal(py_out, jax_out, decimal=6)
```

**Reply by ruili3** [reply]
Thank you, and that's exactly what I do in the [description](https://github.com/google/flax/discussions/2997#:~:text=strides%3D(2%2C%202)%2C-,padding%3D((padding_size%2C%20padding_size)%2C%20(padding_size%2C%20padding_size))%2C,-use_bias%3DFalse%2C%0A%20%20%20%20name). Though the main goal of Flax is not to align with pytorch, can you give some hints on why it can not produce totally the same results using a single conv layer, after explicitly setting the padding?

**Reply by cgarciae** [reply]
Sorry, I missed that part!
Curiously in the script above I am passing the assertion, can you verify that it also works on your machine?
Maybe the pytorch version and the JAX version are running on different devices?
